cli/tests/stake.rs

```diff
@@ -29,9 +33,230 @@ use {
+        nonce_account: None,
+        nonce_authority: 0,
+        memo: None,
+        fee_payer: 0,
+        redelegate_stake_account_pubkey: None,
+    };
+    process_command(&config).unwrap();
+
+    // wait for new epoch
+    wait_for_next_epoch(&rpc_client);
+
+    // `stake_keypair` should now be delegated to `vote_keypair` and fully activated
+    let stake_account = rpc_client.get_account(&stake_keypair.pubkey()).unwrap();
+    let stake_state: StakeState = stake_account.state().unwrap();
+
+    let rent_exempt_reserve = match stake_state {
+        StakeState::Stake(meta, stake) => {
+            assert_eq!(stake.delegation.voter_pubkey, vote_keypair.pubkey());
+            meta.rent_exempt_reserve
+        }
+        _ => panic!("Unexpected stake state!"),
+    };
+
+    assert_eq!(
+        rpc_client
+            .get_stake_activation(stake_keypair.pubkey(), None)
+            .unwrap(),
+        RpcStakeActivation {
+            state: StakeActivationState::Active,
+            active: 50_000_000_000 - rent_exempt_reserve,
+            inactive: 0
+        }
+    );
+    check_balance!(50_000_000_000, &rpc_client, &stake_keypair.pubkey());
+
+    let stake2_keypair = Keypair::new();
+
+    // Add an extra `rent_exempt_reserve` amount into `stake2_keypair` before redelegation to
+    // account for the `rent_exempt_reserve` balance that'll be pealed off the stake during the
+    // redelegation process
+    request_and_confirm_airdrop(
+        &rpc_client,
+        &config,
+        &stake2_keypair.pubkey(),
+        rent_exempt_reserve,
+    )
+    .unwrap();
+
+    // Redelegate to `vote2_keypair` via `stake2_keypair1
+    config.signers = vec![&default_signer, &stake2_keypair];
+    config.command = CliCommand::DelegateStake {
+        stake_account_pubkey: stake_keypair.pubkey(),
+        vote_account_pubkey: vote2_keypair.pubkey(),
+        stake_authority: 0,
+        force: true,
+        sign_only: false,
+        dump_transaction_message: false,
+        blockhash_query: BlockhashQuery::default(),
+        nonce_account: None,
+        nonce_authority: 0,
+        memo: None,
+        fee_payer: 0,
+        redelegate_stake_account_pubkey: Some(stake2_keypair.pubkey()),
+    };
+    process_command(&config).unwrap();
+
+    // check that all the stake, save `rent_exempt_reserve`, have been moved from `stake_keypair`
+    // to `stake2_keypair`
+    check_balance!(rent_exempt_reserve, &rpc_client, &stake_keypair.pubkey());
+    check_balance!(50_000_000_000, &rpc_client, &stake2_keypair.pubkey(),);
+
+    // wait for new epoch
+    wait_for_next_epoch(&rpc_client);
+
+    // `stake_keypair` should now be deactivated
+    assert_eq!(
+        rpc_client
+            .get_stake_activation(stake_keypair.pubkey(), None)
+            .unwrap(),
+        RpcStakeActivation {
+            state: StakeActivationState::Inactive,
+            active: 0,
+            inactive: 50_000_000_000 - rent_exempt_reserve,
```

Comment:
This is very instructive test! We might need to change how RPC fetches the stake activation to show this as just `0` after the deactivation is complete.  If someone is trying to figure out stake distribution on the network just with this, they could get wildly diverging results.

Reply:
I think I see a more general issue with the `getStakeActivation` RPC method. I'll peal this off into another PR then rebase this PR on it

Reply:
302 to https://github.com/solana-labs/solana/pull/26595 for this issue.

I've temporarily cherrypicked that PR into here as well, and accounted for it in this test (now:`inactive: 0` as one would expect)

Reply:
done